mep/accounts/models.py

```diff
@@ -456,43 +454,52 @@ def date_format(cls, value):
+        and precision value.'''
+        # partial date parsing adapted in part from django_partial_date
+        # https://github.com/ktowen/django_partial_date
+        match = self.partial_date_re.match(value)
+        if match:
+            match_info = match.groupdict()
+
+            # turn matched values into numbers for initializing date object
+            date_values = {}
+            date_parts = []
+            for k, v in match_info.items():
+                try:
+                    date_values[k] = int(v)
+                    date_parts.append(k)
+                except (TypeError, ValueError): # value was None or '-'
+                    date_values[k] = self.unknown_year if k == 'year' else 1
+
+            if date_parts == ['day'] or date_parts == ['month'] or date_parts == ['year', 'day']:
+                raise ValidationError('"%s" is not a recognized date.''' % value)
+
+            # determine known date parts based on regex match values
+            # and initialize pecision flags accordingly
+            precision = DatePrecision.from_simple_str('|'.join(date_parts))
+            return (datetime.date(**date_values), precision)
+
+        else:
+            raise ValidationError('"%s" is not a recognized date.''' % value)
+
 
 class Borrow(Event):
     '''Inherited table indicating borrow events'''
     #: :class:`~mep.books.models.Item` that was borrowed;
     #: optional to account for unclear titles
     item = models.ForeignKey(Item, null=True, blank=True)
-    start_date_precision = DatePrecisionField()
-    end_date_precision = DatePrecisionField()
     bought = models.BooleanField(default=False,
         help_text='Item was bought instead of returned')
-
-    def set_partial_start_date(self, value):
-        '''parse a partial date and set :attr:`start_date` and
-        :attr:`start_date_precision` accordingly'''
-        self.start_date, \
-            self.start_date_precision = DatePrecisionField.parse_date(value)
-
-    def set_partial_end_date(self, value):
-        '''parse a partial date and set :attr:`end_date` and
-        :attr:`end_date_precision` accordingly'''
-        self.end_date, \
-            self.end_date_precision = DatePrecisionField.parse_date(value)
-
-    def display_start_date(self):
-        if self.start_date:
-            return self.start_date.strftime(DatePrecisionField.date_format(self.start_date_precision))
-    display_start_date.short_description = 'start_date'
-    display_start_date.admin_order_field = 'start_date'
-
-    def display_end_date(self):
-        if self.end_date:
-            return self.end_date.strftime(DatePrecisionField.date_format(self.end_date_precision))
-    display_end_date.short_description = 'end_date'
-    display_end_date.admin_order_field = 'end_date'
+    start_date_precision = DatePrecisionField(null=True, blank=True)
+    end_date_precision = DatePrecisionField(null=True, blank=True)
+    unknown_year = 1900
```

Comment:
Just a thought in terms of style. I know many django class attributes by convention (though by no requirement) that represent constants are all caps `UNKNOWN YEAR`. I think something similar might improve readability here.

The `PartialDate` property creation is really slick.